ComputerVisionServer/moedx/static/js/index.js

```diff
@@ -0,0 +1,531 @@
+  webcamElement.srcObject = stream;
+  webcamAvailable = true;
+})
+.catch(error => {
+  console.log('navigator.getUserMedia error: ', error);
+  endSession(false);
+  $.alert("Could not connect to webcam. Please load this page on a device with webcam capabilities.");
+});
+
+function restartProcessing() {
+  console.log("restartProcessing() "+currentProtocol+" "+currentEndpoint);
+  if (webSocket != null && webSocket.readyState == WebSocket.OPEN) {
+    webSocket.close(code=1000); // Normal Closure
+  }
+
+  if (currentProtocol == protocolWebSocket) {
+    openWSConnection("wss", window.location.hostname, window.location.port, "/ws" + currentEndpoint);
+  }
+
+  renderData = null;
+  busyProcessing = false;
+  clearTimeout(sessionTimeoutInterval);
+  sessionTimeoutInterval = setTimeout(sessionTimeout, sessionTimeoutMillis);
+  clearInterval(frameInterval);
+  frameInterval = setInterval(processCameraImage, frameMillis);
+  $("#process-onoffswitch").prop("checked", true);
+
+  resetGui();
+}
+
+function resetGui() {
+  if (currentEndpoint == faceDetectionEndpoint) {
+    $("#button-fd").addClass("cv-control-selected");
+  } else if (currentEndpoint == faceRecognitionEndpoint) {
+    $("#button-fr").addClass("cv-control-selected");
+  } else if (currentEndpoint == objectDetectionEndpoint) {
+    $("#button-od").addClass("cv-control-selected");
+  } else if (currentEndpoint == poseDetectionEndpoint) {
+    $("#button-pd").addClass("cv-control-selected");
+  }
+
+  if (currentProtocol == protocolWebSocket) {
+    $("#button-websocket").addClass("cv-control-selected");
+  } else if (currentProtocol == protocolRest) {
+    $("#button-rest").addClass("cv-control-selected");
+  }
+}
+
+function sessionTimeout() {
+  if ($("#process-onoffswitch").prop("checked") ||
+      $("#network-onoffswitch").prop("checked")) {
+        console.log("sessionTimeout() Showing alert dialog.");
+        $.alert("Click an activity button to restart.", "Session Timeout");
+  } else {
+    console.log("sessionTimeout() No activity. Skipping alert dialog.");
+  }
+  $("#network-onoffswitch").prop("checked", false);
+  endSession(true);
+}
+
+function endSession(allowRestart) {
+  console.log("endSession allowRestart="+allowRestart);
+  renderData = null;
+  clearInterval(frameInterval);
+  // clearInterval(networkLatencyInterval);
+  clearTimeout(sessionTimeoutInterval);
+  resetActivityStates() ;
+  $("#process-onoffswitch").prop("checked", false);
+
+  if (!allowRestart) {
+    // Disable all controls except network latency.
+    $(".cv-activity").prop('disabled', true);
+    $("#process-onoffswitch").prop('disabled', true);
+
+    currentEndpoint = null;
+    resetGui();
+  }
+}
+
+restartProcessing();
+getServerCapabilities();
+
+function enableGpuActivities(enabled) {
+  console.log("enableGpuActivities("+enabled+")");
+  $("#button-od").prop('disabled', !enabled);
+  $("#button-pd").prop('disabled', !enabled);
```

Comment:
shouldn't button-fr added to the secret method 😄

Reply:
FR is enabled by default, though any random person doing this process obviously won't be in the training database, and will be recognized as "Unknown". A future improvement could be to add a face training mode like the Android app has.